eslint-webpack-plugin failing the build despite having failOnError set to false

- Operating System: Linux
- Node Version: 14.15.1
- NPM Version: 6.14.8
- webpack Version: 4.44.2
- eslint-webpack-plugin Version: 2.4.0

### Expected Behavior

Show all ESLint errors and warnings, but do not fail the build process. Build should be successful no matter what linting errors are presented.

### Actual Behavior

Despite having failOnError set to false, the build is always failing.

### Code

```js
      plugins: [
          ...
          new ESLintPlugin({
              extensions: 'ts',
              exclude: 'node_modules',
              emitError: true,
              failOnError: false,
              emitWarning: true,
              failOnWarning: false
          })
      ]
```

Failure:

```bash
[09:43:19] webpack built b7ea99f302259f7f4bac in 56785ms
✖ ｢wdm｣: Built at: 11/24/2020 9:43:19 AM

ERROR in 
/[redacted]/foobar.ts
  63:15  error  Multiple spaces found before '('              no-multi-spaces
  63:15  error  Unexpected space before function parentheses  space-before-function-paren
  63:23  error  Missing space before opening brace            space-before-blocks

✖ 3 problems (3 errors, 0 warnings)
  3 errors and 0 warnings potentially fixable with the `--fix` option.

ℹ ｢wdm｣: Failed to compile.
```

I'm also trying an alternative config and it doesn't fail, but it also doesn't show any error or warning:

```js
          new ESLintPlugin({
             extensions: 'ts',
              exclude: 'node_modules',
              emitWarning: true,
              failOnWarning: false
          }),
```

Additionally, setting Webpack's `noEmitOnErrors` to false doesn't affect the build failure:

```js
      optimization: {
          noEmitOnErrors: false
      }
```

I also must keep `.eslintrc` configuring rules as `error` and not `warn` for IDE reporting.

In summary, no matter what config I set, the linting errors are visible and the build is failing (unacceptable) or the build is successful, but nothing about the errors is visible, so at this point the plugin is useless in my case.

--- a/README.md
+++ b/README.md
@@ -188,16 +188,16 @@ Will always return warnings, if set to `true`.
 #### `failOnError`
 
 - Type: `Boolean`
-- Default: `false`
+- Default: `true`
 
-Will cause the module build to fail and not emitted files if there are any errors, if set to `true`.
+Will cause the module build to fail if there are any errors, to disable set to `false`.
 
 #### `failOnWarning`
 
 - Type: `Boolean`
 - Default: `false`
 
-Will cause the module build to fail and not emitted files if there are any warnings, if set to `true`.
+Will cause the module build to fail if there are any warnings, if set to `true`.
 
 #### `quiet`
 

--- a/README.md
+++ b/README.md
@@ -190,14 +190,14 @@ Will always return warnings, if set to `true`.
 - Type: `Boolean`
 - Default: `false`
 
-Will cause the module build to fail if there are any errors, if set to `true`.
+Will cause the module build to fail and not emitted files if there are any errors, if set to `true`.
 
 #### `failOnWarning`
 
 - Type: `Boolean`
 - Default: `false`
 
-Will cause the module build to fail if there are any warnings, if set to `true`.
+Will cause the module build to fail and not emitted files if there are any warnings, if set to `true`.
 
 #### `quiet`
 

--- a/src/ESLintError.js
+++ b/src/ESLintError.js
@@ -1,7 +1,4 @@
-// @ts-ignore
-import WebpackError from 'webpack/lib/WebpackError';
-
-export default class ESLintError extends WebpackError {
+class ESLintError extends Error {
   /**
    * @param {string=} messages
    */
@@ -11,3 +8,5 @@ export default class ESLintError extends WebpackError {
     this.stack = '';
   }
 }
+
+export default ESLintError;

--- a/src/index.js
+++ b/src/index.js
@@ -114,17 +114,23 @@ export class ESLintWebpackPlugin {
       compilation.hooks.succeedModule.tap(ESLINT_PLUGIN, processModule);
 
       // await and interpret results
-      compilation.compiler.hooks.emit.tapPromise(ESLINT_PLUGIN, processResults);
+      compilation.hooks.additionalAssets.tapPromise(
+        ESLINT_PLUGIN,
+        processResults
+      );
 
       async function processResults() {
         const { errors, warnings, generateReportAsset } = await report();
 
-        if (warnings) {
+        if (warnings && !options.failOnWarning) {
           // @ts-ignore
           compilation.warnings.push(warnings);
+        } else if (warnings && options.failOnWarning) {
+          // @ts-ignore
+          compilation.errors.push(warnings);
         }
 
-        if (errors) {
+        if (errors && options.failOnError) {
           // @ts-ignore
           compilation.errors.push(errors);
         }

--- a/src/index.js
+++ b/src/index.js
@@ -112,8 +112,9 @@ export class ESLintWebpackPlugin {
 
       // Gather Files to lint
       compilation.hooks.succeedModule.tap(ESLINT_PLUGIN, processModule);
+
       // await and interpret results
-      compilation.hooks.afterSeal.tapPromise(ESLINT_PLUGIN, processResults);
+      compilation.compiler.hooks.emit.tapPromise(ESLINT_PLUGIN, processResults);
 
       async function processResults() {
         const { errors, warnings, generateReportAsset } = await report();

--- a/src/linter.js
+++ b/src/linter.js
@@ -94,12 +94,6 @@ export default function linter(key, options, compilation) {
       parseResults(options, results)
     );
 
-    if (options.failOnError && errors) {
-      throw errors;
-    } else if (options.failOnWarning && warnings) {
-      throw warnings;
-    }
-
     return {
       errors,
       warnings,

--- a/src/options.js
+++ b/src/options.js
@@ -47,6 +47,7 @@ import schema from './options.json';
 export function getOptions(pluginOptions) {
   const options = {
     extensions: 'js',
+    failOnError: true,
     ...pluginOptions,
     ...(pluginOptions.quiet ? { emitError: true, emitWarning: false } : {}),
   };

--- a/src/options.json
+++ b/src/options.json
@@ -23,7 +23,7 @@
       "anyOf": [{ "type": "string" }, { "type": "array" }]
     },
     "failOnError": {
-      "description": "Will cause the module build to fail if there are any errors, if set to `true`.",
+      "description": "Will cause the module build to fail if there are any errors, to disable set to `false`.",
       "type": "boolean"
     },
     "failOnWarning": {

--- a/test/fail-on-error.test.js
+++ b/test/fail-on-error.test.js
@@ -4,8 +4,9 @@ describe('fail on error', () => {
   it('should emits errors', (done) => {
     const compiler = pack('error', { failOnError: true });
 
-    compiler.run((err) => {
-      expect(err.message).toContain('error.js');
+    compiler.run((err, stats) => {
+      expect(err).toBeNull();
+      expect(stats.hasErrors()).toBe(true);
       done();
     });
   });

--- a/test/fail-on-warning.test.js
+++ b/test/fail-on-warning.test.js
@@ -4,8 +4,9 @@ describe('fail on warning', () => {
   it('should emits errors', (done) => {
     const compiler = pack('warn', { failOnWarning: true });
 
-    compiler.run((err) => {
-      expect(err.message).toContain('warn.js');
+    compiler.run((err, stats) => {
+      expect(err).toBeNull();
+      expect(stats.hasErrors()).toBe(true);
       done();
     });
   });

